### ISSUE TITLE
Decrease the number of Kubernetes resources we cache

### ISSUE THREAD

**Issue by ezimanyi** [spinnaker/spinnaker#5905]
### Overview

The Kubernetes provider exhaustively adds the full manifest for any kind that it can read to the cache.  In most cases, the cost of this caching outweighs the benefit, and contributes to difficultly scaling the Kubernetes provider to large deployments.

The goal here is to re-think what we cache, and only cache things when it actually does lead to better performance than making live lookups against the cluster.

### Details

* Currently, Clouddriver pre-caches every possible resource (everything Spinnaker has permission to list, every Clouddriver built-in kind, every config-registered custom resource).
* We should only cache the minimum resource types necessary for rendering resources by cluster in Deck’s infrastructure view, and selecting cluster targets dynamically.
  * Deck’s infrastructure view only renders resource types that map to Spinnaker kinds.
  * Dynamic target selection is currently available for all built-in and custom-registered resource types.
  * Static target selection (where resource names are pre-populated for Enable, Disable, Patch, and Delete stages) should rely on live calls instead of the cache.
* For every other type of resource, it’s not necessary to cache unless/until we are trying to apply it. All of the resources with handlers for Spinnaker-managed versioning or rollout strategies are built-in kinds.
* It should not be necessary to register a CRD in an account’s customResources block in order to apply it successfully.

This will depend on first [removing live manifest calls](https://github.com/spinnaker/spinnaker/issues/5906) (and having all deploy stages do live queries) as otherwise kinds that are removed from the cache will not be able to be deployed.

**Comment by ncknt** [comment]
> All of the resources with handlers for Spinnaker-managed versioning or rollout strategies are built-in kinds
CRDs (or event standard kinds) can be changed via plugins.

**Comment by ezimanyi** [comment]
@ncknt : Do you have some more details about how these can be changed via plugins?  What changes are possible, and how does a user configure them?

Ultimately the goal here is that the default caching strategy should be more sensible, and only cache things where the benefit of caching outweighs the cost.  In the case of CRDs, we're caching the full text of all deployed CRDs, when in general the only queries we do against the cache are "get me the text of this CRD".

The changes here feel like something that would benefit all OSS users; I'd rather not need users to install a plugin to get scaleable caching performance.

**Comment by ncknt** [comment]
You can load new `KubernetesHandler` via a plugin and start deploying versioned CRDs (for example). It doesn't really change anything other than perhaps not just built-in kinds have their version managed by Spinnaker.

> Static target selection (where resource names are pre-populated for Enable, Disable, Patch, and Delete stages) should rely on live calls instead of the cache.

Did you mean dynamic here? Isn't it for these stages that the name of the manifest is dynamically retrieved?

**Comment by ezimanyi** [comment]
@ncknt : Yes, we'd definitely keep compatibility with the current ability to override a `KubernetesHandler`.  This would be a lower-level change, where we'd no longer automatically fill the cache with the full manifest for every kind (whether built-in or CRD).  The cache would instead focus on supporting queries that are expensive to do without the cache.  If there are queries that CRD handlers can do that require the cache, then we'd need to make sure we can still support those using the cache...but we'd stop automatically caching everything.

> Did you mean dynamic here? Isn't it for these stages that the name of the manifest is dynamically retrieved?

This actually does mean static.  The line above was referring to dynamic target selection...but for static target selection, the UI loads a list of static targets.  This list is currently supported by the cache, and so we'd need to figure out how to either (1) load the list without the cache (2) support the list in the cache, (3) eliminate the pre-populated list.

**Comment by ncknt** [comment]
I see. It's dynamically static :) Thanks for the explanation.

**Comment by jasonmcintosh** [comment]
This... has actually been completed and configurable on various accounts now for a while.  Further in later releases, caching of k8s operations can be completely disabled while still allowing deployments to operate fine, as deploys do not depend on the cache for k8s operations.  That said, there are options to enable/disable this selectively.  Closing this for now, and will have other features in the future.

https://github.com/spinnaker/spinnaker/blob/main/clouddriver/clouddriver-kubernetes/src/main/java/com/netflix/spinnaker/clouddriver/kubernetes/config/KubernetesConfigurationProperties.java#L96 please see this for more information
